hail/src/test/scala/is/hail/io/compress/BGzipCodecSuite.scala

```diff
@@ -114,4 +115,74 @@ class BGzipCodecSuite extends SparkSuite {
     }
     p.check()
   }
+
+  @Test def testVirtualSeek() {
+    // real offsets of the start of some blocks, paired with the offset to the next block
+    val blockStarts = Array[(Long, Long)]((0, 14653), (69140, 82949), (133703, 146664), (181362, 192983 /* end of file */))
+    // magic number determined by indexing the target file
+    val maxBlockSize = 65280
+    // magic number determined by counting bytes from sample.vcf from uncompBlockStarts(-1) to the end of the file
+    val lastBlockLen = 55936
+    // offsets into the uncompressed file
+    val uncompBlockStarts = Array[Int](0, 326400, 652800, 913920)
+
+    val uncompPath = "src/test/resources/sample.vcf"
+    val compPath = "src/test/resources/sample.vcf.gz"
+
+    val uncompHPath = new hd.fs.Path(uncompPath)
+    val compHPath = new hd.fs.Path(compPath)
+
+    val fs = uncompHPath.getFileSystem(hadoopConf)
+    val compIS = fs.open(compHPath)
+    using (fs.open(uncompHPath)) { uncompIS => using (new BGzipInputStream(compIS)) { decompIS =>
+
+      for (((cOff, nOff), uOff) <- blockStarts.zip(uncompBlockStarts);
+           e <- Seq(0, 1024, maxBlockSize)) {
+        val decompData = new Array[Byte](100)
+        val uncompData = new Array[Byte](100)
+        val extra = if (cOff == blockStarts(blockStarts.size-1)._1 && e == maxBlockSize)
+            lastBlockLen
+          else
+            e
+        val vOff = BlockCompressedFilePointerUtil.makeFilePointer(cOff, extra)
+
+        decompIS.virtualSeek(vOff)
+        assert(decompIS.getVirtualOffset() == vOff);
+        uncompIS.seek(uOff + extra)
+
+        val decompRead = decompIS.read(decompData)
+        val uncompRead = uncompIS.read(uncompData)
+
+        assert(decompRead == uncompRead, s"""compressed offset: ${ cOff }
+          |decomp bytes read: ${ decompRead }
+          |uncomp bytes read: ${ uncompRead }\n""".stripMargin)
+        assert(decompData sameElements uncompData, s"data differs for compressed offset: ${ cOff }")
+        val expectedVirtualOffset = if (e != maxBlockSize)
+            BlockCompressedFilePointerUtil.makeFilePointer(cOff, extra + decompRead)
+          else
+            BlockCompressedFilePointerUtil.makeFilePointer(nOff, if (decompRead == -1) 0 else decompRead)
+        assert(expectedVirtualOffset == decompIS.getVirtualOffset())
+      }
+
+      // Trying to seek to the end of a block should succeed
+      val vOff = BlockCompressedFilePointerUtil.makeFilePointer(blockStarts(1)._1, maxBlockSize);
+      decompIS.virtualSeek(vOff)
+
+      // Trying to seek past the end of a block should fail
+      intercept[java.io.IOException] {
+        val vOff = BlockCompressedFilePointerUtil.makeFilePointer(blockStarts(0)._1, maxBlockSize + 1);
+        decompIS.virtualSeek(vOff)
+      }
+
+      // trying to seek to exactly end of file should succeed
+      val lastOffset = BlockCompressedFilePointerUtil.makeFilePointer(blockStarts(blockStarts.size - 1)._1, lastBlockLen)
```

Comment:
As per our discussion, the end of file is (endOfFilePos, 0).  This shouldn't pass.

Reply:
That was unclear to me. I thought we were allowing both (blockNPos, blockNUncompLength) and (blockNPlus1Pos, 0) as valid addresses. If this is the case, then eof is both (endOfFilePos, 0) and (penultimateBlockPos, penultimateBlockLength), otherwise, eof is just (endOfFilePos, 0).